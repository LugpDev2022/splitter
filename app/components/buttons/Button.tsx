interface ButtonProps {
  active?: boolean;
  children: React.ReactNode;
  className?: string;
  disabled?: boolean;
  onClick: () => void;
}

const Button: React.FC<ButtonProps> = ({
  active,
  children,
  className,
  disabled,
  onClick,
}) => {
  return (
    <button
      onClick={onClick}
      disabled={disabled}
      className={`
      ${active ? 'bg-teal-500' : 'bg-cyan-900'}
      ${active ? 'text-cyan-900' : ' text-white'}
      ${className ? className : ''}
      disabled:bg-[#0d686d]
      disabled:hover:bg-[#0d686d]
      disabled:hover:text-[#00494d]
      disabled:text-[#00494d]
      duration-300
      font-bold
      hover:bg-teal-200
      hover:text-cyan-900
      p-4
      rounded-lg
      text-xl
      transition
    `}
    >
      {children}
    </button>
  );
};

export default Button;
