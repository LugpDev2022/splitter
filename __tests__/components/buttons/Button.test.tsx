import { fireEvent, render, screen } from '@testing-library/react';
import Button from '@/app/components/buttons/Button';

describe('Tests on <Button />', () => {
  const onClick = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should have normal button styles', () => {
    const { container } = render(
      <Button onClick={onClick}>Test Button</Button>
    );

    const button = screen.getByRole('button');
    const classNames = button.getAttribute('class');

    expect(classNames).toContain('text-white');
    expect(classNames).toContain('bg-cyan-900');
    expect(container).toMatchSnapshot();
  });

  test('should have active button styles', () => {
    const { container } = render(
      <Button onClick={onClick} active>
        Test Button
      </Button>
    );

    const button = screen.getByRole('button');
    const classNames = button.getAttribute('class');

    expect(classNames).toContain('text-cyan-900');
    expect(classNames).toContain('bg-teal-500');
    expect(container).toMatchSnapshot();
  });

  test('should have disabled styles', () => {
    render(<Button onClick={onClick}>Test Button</Button>);

    const button = screen.getByRole('button');
    const classNames = button.getAttribute('class');

    expect(classNames).toContain('disabled:bg-[#0d686d]');
    expect(classNames).toContain('disabled:hover:bg-[#0d686d]');
    expect(classNames).toContain('disabled:hover:text-[#00494d]');
    expect(classNames).toContain('disabled:text-[#00494d]');
  });

  test('should have the given css classes', () => {
    const testClassNames = 'testClass anotherTestClass';

    render(
      <Button onClick={onClick} className={testClassNames}>
        Test Button
      </Button>
    );

    const button = screen.getByRole('button');
    const classNames = button.getAttribute('class');
    expect(classNames).toContain(testClassNames);
  });

  test('should render the children', () => {
    const buttonLabel = 'Test Button Label';

    render(<Button onClick={onClick}>{buttonLabel}</Button>);

    expect(screen.getByText(buttonLabel)).toBeTruthy();
  });

  test('should disable the button', () => {
    render(
      <Button onClick={onClick} disabled>
        Test button
      </Button>
    );

    const button: HTMLButtonElement = screen.getByRole('button');
    expect(button.disabled).toBeTruthy();
  });

  test('should call the onClick function', () => {
    render(<Button onClick={onClick}>Test button</Button>);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('should not call the onClick function when disabled', () => {
    render(
      <Button onClick={onClick} disabled>
        Test button
      </Button>
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(onClick).not.toHaveBeenCalled();
  });
});
