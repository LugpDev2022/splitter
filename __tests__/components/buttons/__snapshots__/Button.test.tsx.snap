// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <Button /> should have active button styles 1`] = `
<div>
  <button
    class="
    bg-teal-500
    text-cyan-900
    
    disabled:bg-[#0d686d]
    disabled:hover:bg-[#0d686d]
    disabled:hover:text-[#00494d]
    disabled:text-[#00494d]
    duration-300
    font-bold
    hover:bg-teal-200
    hover:text-cyan-900
    p-4
    rounded-lg
    text-xl
    transition
  "
  >
    Test Button
  </button>
</div>
`;

exports[`Tests on <Button /> should have normal button styles 1`] = `
<div>
  <button
    class="
    bg-cyan-900
     text-white
    
    disabled:bg-[#0d686d]
    disabled:hover:bg-[#0d686d]
    disabled:hover:text-[#00494d]
    disabled:text-[#00494d]
    duration-300
    font-bold
    hover:bg-teal-200
    hover:text-cyan-900
    p-4
    rounded-lg
    text-xl
    transition
  "
  >
    Test Button
  </button>
</div>
`;
